flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/OperatorCoordinatorHolder.java

```diff
@@ -238,8 +241,11 @@ public void notifyCheckpointAborted(long checkpointId) {
     @Override
     public void resetToCheckpoint(long checkpointId, @Nullable byte[] checkpointData)
             throws Exception {
-        // ideally we would like to check this here, however this method is called early during
-        // execution graph construction, before the main thread executor is set
+        // the first time this method is called is early during execution graph construction,
+        // before the main thread executor is set. hence this conditional check.
+        if (mainThreadExecutor != null) {
+            mainThreadExecutor.assertRunningInMainThread();
+        }
```

Comment:
I guess "known" does not means "run on". The construction of whole RPC endpoint is not run in the main thread, so the assertion will fail in my expectation.

Reply:
Let's definitely look at this in a follow-up. It'd be super happy to get rid of the lazy initialization, if possible.

But that means that the main thread executor would need to be already used for the call that restores the initial savepoint. We should double check whether that is the case.

Reply:
I think what needs to change from the `JobMasters` perspective is that the `Scheduler` has to be created when `JobMaster.onStart` is called. Of course, if the scheduler creates the `EG` synchronously, then this will block the main thread which shouldn't be a huge problem, though. In any case, let's tackle this as a potential follow up.